render/pixman/renderer.c

```diff
@@ -0,0 +1,339 @@
+
+	return texture->format == WL_SHM_FORMAT_XRGB8888;
+}
+
+static bool texture_write_pixels(struct wlr_texture *wlr_texture,
+		uint32_t stride, uint32_t width, uint32_t height,
+		uint32_t src_x, uint32_t src_y, uint32_t dst_x, uint32_t dst_y,
+		const void *data) {
+	assert(false && "todo texture_write_pixels");
+	return false;
+}
+
+static bool texture_to_dmabuf(struct wlr_texture *wlr_texture,
+		struct wlr_dmabuf_attributes *attribs) {
+	assert(false && "todo texture_to_dmabuf");
+	return false;
+}
+
+static void texture_destroy(struct wlr_texture *wlr_texture) {
+	if (wlr_texture == NULL) {
+		return;
+	}
+	struct wlr_pixman_texture *texture = get_texture(wlr_texture);
+
+	pixman_image_unref(texture->image);
+	free(texture);
+}
+
+static const struct wlr_texture_impl texture_impl = {
+	.is_opaque = texture_is_opaque,
+	.write_pixels = texture_write_pixels,
+	.to_dmabuf = texture_to_dmabuf,
+	.destroy = texture_destroy,
+};
+
+static uint32_t get_pixman_format_from_wl(enum wl_shm_format wl_fmt) {
+	uint32_t fmt = 0;
+	switch(wl_fmt) {
+	case WL_SHM_FORMAT_ARGB8888:
+	#if WLR_BIG_ENDIAN
+		fmt = PIXMAN_a8r8g8b8;
+	#elif WLR_LITTLE_ENDIAN
+		fmt = PIXMAN_b8g8r8a8;
+	#else
+	#error pixman endian error
+	#endif
+		break;
+	case WL_SHM_FORMAT_XRGB8888:
+	#if WLR_BIG_ENDIAN
+		fmt = PIXMAN_x8r8g8b8;
+	#elif WLR_LITTLE_ENDIAN
+		fmt = PIXMAN_b8g8r8x8;
+	#else
+	#error pixman endian error
+	#endif
+		break;
+	default:
+		// TODO: wl_shm_format to string?
+		wlr_log(WLR_ERROR, "Unsupported wl_shm_format");
+		break;
+	}
+
+	return fmt;
+}
+
+struct wlr_pixman_texture *pixman_create_texture(
+		struct wlr_texture *wlr_texture, struct wlr_pixman_renderer *renderer);
+
+static void handle_destroy_buffer(struct wl_listener *listener, void *data) {
+	struct wlr_pixman_buffer *buffer =
+		wl_container_of(listener, buffer, buffer_destroy);
+	wl_list_remove(&buffer->link);
+	wl_list_remove(&buffer->buffer_destroy.link);
+
+	pixman_image_unref(buffer->image);
+
+	free(buffer);
+}
+
+static struct wlr_pixman_buffer *create_buffer(
+		struct wlr_pixman_renderer *renderer, struct wlr_buffer *wlr_buffer) {
+	struct wlr_pixman_buffer *buffer = calloc(1, sizeof(*buffer));
+	if (buffer == NULL) {
+		wlr_log_errno(WLR_ERROR, "Allocation failed");
+		return NULL;
+	}
+	buffer->buffer = wlr_buffer;
+	buffer->renderer = renderer;
+
+	struct wlr_dmabuf_attributes dmabuf = {0};
+	if (!wlr_buffer_get_dmabuf(wlr_buffer, &dmabuf)) {
+		goto error_buffer;
+	}
+
+	uint32_t format;
+	switch (dmabuf.format) {
+	case DRM_FORMAT_XRGB8888:
+		format = PIXMAN_x8r8g8b8;
+		break;
+	case DRM_FORMAT_ARGB8888:
+		format = PIXMAN_a8r8g8b8;
+		break;
+	default:
+		wlr_log(WLR_ERROR, "Unsupported DRM format 0x%x\n", dmabuf.format);
+		goto error_buffer;
+	}
+
+	void *data = NULL;
+	size_t size = 0;
+	if (!wlr_buffer_get_data_ptr(wlr_buffer, &data, &size)) {
+		wlr_log(WLR_ERROR, "Failed to get buffer data");
+		goto error_buffer;
+	}
+
+	uint32_t stride = size / wlr_buffer->height;
```

Comment:
Hm. This sounds a little fragile, but is probably good enough for now. Maybe add an assert to check that the size is divisible by the height.

Reply:
Maybe if we have a way to get it directly from the buffer itself, like adding a func to the impl `get_stride` ?

Reply:
We can't have a generic `get_stride` function, because some buffers don't have a single stride. Multi-planar buffers may have up to 4 different strides.

Maybe we can just add the stride as an out-param of `get_data_ptr`.

Reply:
`bool wlr_buffer_get_stride` and return `false` if its not supported?

Reply:
Hm, I'm not sure. I think I'd prefer the assumption "there is only a single stride" to be tied to `get_shm` and `get_data_ptr`.

Reply:
BTW, we could just replace the size out-param with stride. There's non need to know the exact size when you know you can access the buffer up to `stride * height`.

Reply:
Works for me, the only place where I use the size is to calculate the stride. Shall I do the edit in this branch or in `shm-allocator`?

Reply:
As you prefer. I'll backport to my branch if you change it here.